src/compiler/path.ts

```diff
@@ -461,12 +461,15 @@ namespace ts {
      */
     export function reducePathComponents(components: readonly string[]) {
         if (!some(components)) return [];

```

Comment:
I just looked up usage of `reducePathComponents` and in all places it is passed value of `getPathComponents`. `reducePathComponents` is internal so do you think we could operate on same array instead of creating new one in all cases? And we could totally document that behaviour and still be ok.

Reply:
I like the idea, but won't that require a lot of mid-array deletions?

Reply:
If you provide an alternate implementation, I'll be happy to test it against the real repro.

Reply:
Something like this?
```ts
    export function reducePathComponents(components: string[]) {
        let reducedNext = 1;
        for (let i = reducedNext; i < components.length; i++) {
            const component = components[i];
            if (!component) continue;
            if (component === ".") continue;
            if (component === "..") {
                if (reducedNext > 1) {
                    if (components[reducedNext - 1] !== "..") {
                        reducedNext--;
                        continue;
                    }
                }
                else if (components[0]) continue;
            }
            if (i !== reducedNext) components[reducedNext] = component;
            reducedNext++;
        }
        components.length = reducedNext;
        return components;
    }

```

Reply:
That's much nicer than what I was trying to do. :)  Still no effect though.  I think the issue is that it's dwarfed by the cost of the conversions from string to array and back.  I'd definitely take your change over mine, but I'm not sure either is worthwhile.